lib/zone.ts

```diff
@@ -1232,5 +1253,151 @@ const Zone: ZoneType = (function(global: any) {
 
   // This is not part of public API, but it is usefull for tests, so we expose it.
   Promise[Zone.__symbol__('uncaughtPromiseErrors')] = _uncaughtPromiseErrors;
+
+  /*
+   * This code patches Error so that:
+   *   - It ignores un-needed stack frames.
+   *   - It Shows the associated Zone for reach frame.
+   */
+
+  enum FrameType {
+    /// Skip this frame when printing out stack
+    blacklist,
+    /// This frame marks zone transition
+    trasition
+  };
+  const NativeError = global[__symbol__('Error')] = global.Error;
+  // Store the frames which should be removed from the stack frames
+  const blacklistedStackFrames: {[frame: string]:FrameType} = {};
+  // We must find the frame where Error was created, otherwise we assume we don't understand stack
+  let zoneAwareFrame: string;
+  global.Error = ZoneAwareError;
+  // How should the stack frames be parsed.
+  let frameParserStrategy = null;
+
+
+  /**
+   * This is ZoneAwareError which processes the stack frame and cleans up extra frames as well as
+   * adds zone information to it.
+   */
+  function ZoneAwareError() {
+    // Create an Error.
+    let error: Error = NativeError.apply(this, arguments);
+
+    // Save original stack trace
+    error.originalStack = error.stack || '';
+
+    // Process the stack trace and rewrite the frames.
+    if (ZoneAwareError['rewriteStack']) {
```

Comment:
Since generating errors isn't a performance concern (we don't throw millions of errors), have you considered just creating a new stack instead of modifying the existing one in place?  The code will be easier for follow.